### PR TITLE
fix(jobs): the five follow-ups the job contract left open

### LINKED ISSUE #428: lint: //nolint:forbidigo is unusable tree-wide (two configs, two pattern sets, nolintlint)
## What

`.golangci.yml` and `.golangci.strict.yml` both enable `forbidigo`, with
**different** `forbid:` pattern sets, and both enable `nolintlint`. A
`//nolint:forbidigo` directive written for one pass is therefore reported as an
*unused* nolint directive by the other, and `make check` goes red — so the
in-source waiver mechanism is unusable anywhere in the tree for this one linter.

The workaround is already in the repo and already documented as a workaround.
`.golangci.yml` excludes the one sanctioned `river.AddWorker` call site by
**path** rather than by a directive, and says why:

> `jobregistry.go` holds the ONE sanctioned `river.AddWorker` … Excluded by path
> rather than by a `//nolint`, because the strict pass enables `forbidigo` with
> different patterns and its `nolintlint` would then read the directive as unused.

## Why it matters

A path exclusion is coarser than the thing it stands in for. It exempts the
whole file from the linter rather than one line, so a second, genuinely
unsanctioned direct registration added to that same file later would not be
caught. Every future `forbidigo` rule inherits the same limitation: the only
available waiver is "exempt a file".

It is also a trap rather than a defect anyone hits once — the failure appears in
a linter unrelated to the rule being waived, on a line that looks correct.

## Options (not a decision)

- Give the two configs one shared `forbidigo` pattern set, so a directive is
  valid under both passes.
- Disable `nolintlint`'s unused-directive check for `forbidigo` specifically.
- Keep path exclusions and state the limitation in `.golangci.yml` as the
  standing rule rather than as a note on one entry.

## Scope

Repo-level linting configuration. Found while landing the job-kind contract
(`backend/api/jobs.yaml`), but not an artefact of it — nothing in that change
depends on which option is taken.

### PR DESCRIPTION
**Pull request by tinnhlt-gra**
Closes #428, #429, #430, #431, #432 — every follow-up filed from #446.

## What changes

**#432 — a dispatcher enqueued by hand runs the whole fleet.** One tenant finishing a capture backfill enqueued the `capture_digest` **dispatcher**, which fans one job out per workspace across the installation. The intent at that call site is local; the effect scaled with the number of tenants running backfills. It now enqueues the child kind for the one workspace it already knows, through `oneOffChildOpts` — declared queue and attempt cap, and deliberately *neither* the sweep tag (a job one tenant asked for is not one workspace's share of a fleet pass) nor the active-state uniqueness (the event fired *because* new rows landed; a pass already running may have read the workspace before they did).

Closed as a class, not at the call site: `TestNoScheduledDispatcherIsEnqueuedByHand` refuses a dispatcher args value constructed anywhere but `periodicFor`, in all three spellings (composite literal, `var`, `new(T)`). The one exception is **derived** from the contract rather than listed — a kind whose declared cadence is `on_demand` has no clock, so a call site is the only thing that can place it.

**#431 — a failed connection the workspace pair cannot see.** `margince_sweep_workspaces_*` counts distinct workspaces, and four dispatchers fan out below that grain. A workspace holding two connections produces two children per pass; if the broken one failed *before* the healthy one succeeded, the most recent child is the successful one and the pair reports zero failures.

Making the existing pair per-child would break what it is for — River resolves a uniqueness conflict by updating the existing row, so there is no "last pass" in `river_job` and any batch-keyed reading under-reports a dispatcher retried mid-fleet. So this is a **second** pair at the declared grain: `margince_sweep_units_total` / `_failed`, grouped on the args key the declared `fan_out_unit` names, with kind and key passed as two aligned arrays joined against `river_job` (neither reaches the statement as text).

The two families **overlap rather than partition** — a per-connection kind is reported by both, at two grains — which is stated in the docs together with *never sum them*.

Nothing here is hand-maintained: `FanOutUnits` walks the declared edges backwards, `ArgsKey` is a switch over the unit enum gated by a test that reads the units the **contract** declares, and a census arm holds every fan-out child to actually *writing* that key — checked against `encoding/json`, because a Go field named `ConnectionID` satisfies a name-level walk while the column holds `connection_id`.

**#430 — `cmd/worker` published nothing about itself.** `--observe-addr` (env `MARGINCE_OBSERVE_ADDR`, default **off**) serves `/healthz`, `/readyz` and `/metrics`. It carries only what is process-local — Go runtime, this process's pool, this process's relay counter — and re-serves no job-table gauge and no outbox backlog: those stay a single fleet-wide reading on the api, and a test asserts both directions by name. `/readyz` probes `boot`, `postgres` and `redis`; the boot gate exists because the listener comes up *before* the lanes and the runner on purpose, and without it a rollout could retire the last working replica for one that had not yet picked up a job.

**#428 — the waiver that could not be written.** `forbidigo` was enabled in both lint configs with different pattern sets, so an in-source waiver written for one was reported as *unused* by the other's `nolintlint`, failing the build in a linter unrelated to the rule being waived. It now lives in `.golangci.yml` alone, held there by a gate. Its two extra rules became repo-wide, which cost one real fix: the split-update handler answered a marshal failure with a `text/plain` `http.Error` on a surface where every other refusal is an RFC 7807 document.

**#429** — dropped the `STATUS.md` pointer from `replayscope.go`'s type comment.

## Gate evidence

Each gate was watched to fail before being trusted.

**The dispatcher gate, against the reverted fix:**
```
--- FAIL: TestNoScheduledDispatcherIsEnqueuedByHand
    capturejobs.go:336:34: CaptureDigestArgs (capture_digest) is constructed outside periodicFor.
    It is a dispatcher with a declared cadence, so enqueueing it fans one job out per workspace
    across the whole installation
```
And against a `var` declaration, which the first version of the walk could not see: `capturejobs.go:264:5: CaptureDigestArgs (capture_digest) is constructed outside periodicFor`.

**The unit-key census arm, with `ArgsKey` repointed at a key nothing writes:**
```
--- FAIL: TestJobCensusMatchesTheContract
    capture_sync is fanned out per connection but its rows carry no "conn_id" key
    (they carry [connection_id provider workspace_id]) — the sweep gauges group on that key,
    so every child of this kind would be dropped from the count rather than reported as a failure
```

**The #428 trap, reproduced and then removed.** With `forbidigo` re-added to the strict config under a different pattern set:
```
internal/compose/jobregistry.go:173:2: directive `//nolint:forbidigo // the ONE sanctioned
registration: …` is unused for linter "forbidigo" (nolintlint)
```
With the fix in place both passes report `0 issues.`

**The backfill-digest integration test, against the old behaviour:**
```
--- FAIL: TestBackfillCompletionBuildsTheDigest
    capture_digest rows went 1 -> 2 across the backfill; a completion must enqueue the CHILD kind
    0 untagged capture_digest_workspace row(s), want exactly 1
```

## UAT — live stack, two independent runs

Run against `make dev` (api :18080, worker observe surface :19101). Both agents reported PASS on every item.

**#430 — the worker surface.** `/healthz` → `200 ok`. `/readyz` → `200 ready`. `/metrics` serves exactly six families and no more:
```
margince_process_goroutines 83
margince_process_heap_bytes 26671128
margince_process_heap_sys_bytes 36339712
margince_process_gc_cycles_total 1
margince_relay_published_total 0
margince_pgxpool_conns{state="acquired"} 0
margince_pgxpool_conns{state="idle"} 7
margince_pgxpool_conns{state="total"} 7
margince_pgxpool_conns{state="max"} 16
```
`margince_job_queue_depth`, `margince_job_declared_info`, `margince_sweep_workspaces_total`, `margince_sweep_units_total` and `margince_outbox_unpublished` are absent as raw substrings — not even a HELP line mentions them. Every series has a `# HELP` and a `# TYPE`.

Proven process-local rather than the same reading twice, across three consecutive scrape rounds:
```
worker: goroutines 83→84  heap 28116632→28131456  pgxpool total 7
api   : goroutines  9→ 9  heap 23993016→24125080  pgxpool total 3
```
Two distinct PIDs behind the two ports. The live worker's command line carries **no** `--observe-addr` flag — the surface came on purely from the env var, so that path is proven end to end, and `--help` shows no default (Go prints one only for a non-zero value).

**#431 — the masking, reproduced in the live database.** One workspace, two connections, kind `telegram_poll`, both `sweep`-tagged; the broken one discarded 2h ago, the healthy one completed 1h ago.

Before:
```
# TYPE margince_sweep_units_total gauge      (header present, no series — idle fleet)
# TYPE margince_sweep_units_failed gauge
```
After the inserts — this is the issue, on the wire:
```
margince_sweep_workspaces_total{sweep="telegram_poll"} 1
margince_sweep_workspaces_failed{sweep="telegram_poll"} 0
margince_sweep_units_total{sweep="telegram_poll",unit="connection"} 2
margince_sweep_units_failed{sweep="telegram_poll",unit="connection"} 1
```
Then a tagged `capture_digest_workspace` row was added for the same workspace and appeared in `margince_sweep_workspaces_*` while `margince_sweep_units_*` still contained only the two `telegram_poll` lines — proving the exclusion is a grain **selection**, not an absence of data. Every seeded row was deleted and the series returned to their prior state.

## Review provenance

Two independent whole-branch reviews (Fable and Codex) before UAT. Between them they found the comment-asserts-a-false-invariant class **five** times — three introduced by this branch — plus three real gaps in the worker listener: `/readyz` answering ready before the lanes and runner started, no read/write/idle timeouts on an unauthenticated listener, and a reserve-close-rebind port race in the tests. All fixed; the listener now returns the address it bound, so `:0` works and "off bound nothing" is a value rather than an absence inferred from a call that did not error.

Two gates were **widened** rather than patched as a result: the dispatcher walk (three spellings, not one) and the baseline lint config, which now declares the `integration`/`livesmoke` tags — moving `forbidigo` out of the strict config had silently dropped the tagged harnesses out of every pass. Zero new findings from that widening, and a test holds it.

UAT then found three more the tests could not: an empty `ai:` line in the worker's readiness body, no browser-facing headers on the new surface, and a real dashboard hazard — the `sweep` label is the **child** kind, so a join to `margince_job_declared_info` lands on an entry carrying no `fan_out_unit`. All three fixed or documented.

**Bot round.** CodeRabbit and cubic then found 15 more between them. Ten were real and are fixed — including three genuine defects in this branch's own new code that neither earlier review nor UAT caught:

- `FanOutUnits` walked an **unordered map and overwrote on collision**. Nothing in the contract stopped two dispatchers naming one child with different units, and the winner would then be whichever the range reached last — a grain nobody chose, and one that could differ between processes. Closed at *generation* now, where a contract can still be edited; two edges sharing a child with the same unit stay legal, because the rule is about disagreement, not sharing.
- The unit read **dropped `statsBySweep`'s workspace predicate**, so a malformed row carrying a connection id and an empty `workspace_id` would credit a pass with coverage belonging to no tenant.
- `/readyz` **stayed green through the whole shutdown**. The listener is stopped last on purpose so the drain stays observable, which meant a terminating replica kept advertising itself while the runner and lanes were being put down. `bootGate` answers no at both ends of the process's life now, with `draining()` deferred after `complete()` so LIFO puts it ahead of the runner stop — 200 before, 503 after, proven in the integration lane.

Plus one unfalsifiable test of mine (the `ai:` omission asserted over a 503 body that never had the line), one env-dependent one (`--dsn` backs its default with `MARGINCE_DSN`), and four documentation and error-message corrections.

Four were declined with reasoning rather than silence, and two of those became follow-up issues: #460 (both sweep pairs carry a `_total` suffix on a gauge — a convention point that applies to all four families or to none, and a breaking dashboard change either way) and #461 (`/metrics` and `/readyz` discard response-write failures while the job section does not — pre-existing, and the split is principled but invisible). Every review thread is answered and resolved.

## What is deliberately NOT claimed

- **Both `_failed` halves see only what River sees.** A kind declaring `fault.nil_after_logging` — `capture_sync` and `voice_build`, whose retry cadence belongs to their own sidecar — records its own failure and completes green, so a handled failure does not reach either pair. This is a property of the sweep reading as a whole rather than of the new half, and it is now stated where an operator reads the number.
- **The dispatcher-enqueue gate walks this package's own top-level product sources.** A dispatcher value built in a subpackage, in a `cmd`, or returned from a helper is outside it. Stated in the test rather than left implied; the enqueue door itself is held separately by the generated closed union.
- **`unit="build"` is unexercised on the wire** — no live build data existed during UAT. The family and label rendering are unit-tested.

## Coverage

**84.1%** of this branch's new statements (246 statements, 207 covered), measured across the unit and integration lanes. Two files sit under the bar and neither is new ground: `cmd/worker/main.go` at 42.9% is `run()`, which needs a full boot to reach and was untested before this branch, and `internal/compose/agentsplit.go` at 66.7% is the split handler's staging path, which has no handler test at all — the lines this branch changed in it are covered.

## Gates run locally

`make check-backend`, `make check-fe`, `make craft-static`, `make test-integration` (`OK: integration passed with 0 skips`), and both `golangci-lint` passes (`0 issues.` each) — all after merging `origin/main`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Prevents accidental fleet-wide dispatches, adds per-unit sweep metrics, and gives `cmd/worker` its own health, readiness, and metrics surface. Also hardens readiness during shutdown, validates the job contract for unit consistency, and fixes docs and error shapes.

- **New Features**
  - Per-unit sweep metrics: `margince_sweep_units_total` and `_failed`, grouped by declared `unit` (e.g., `connection`, `build`). Docs note overlap with workspace pair.
  - `cmd/worker` observe surface: `--observe-addr` (env `MARGINCE_OBSERVE_ADDR`) serves `/healthz`, `/readyz` (boot/postgres/redis), and `/metrics` with process-local gauges only.

- **Bug Fixes**
  - Prevent fleet-wide dispatches: backfill completion enqueues the workspace child via `oneOffChildOpts` (no sweep tag or active-unique); gate rejects hand-enqueueing scheduled dispatchers outside `periodicFor` (except `on_demand`).
  - Metrics correctness and contract validation: enforce workspace scoping in unit reads; refuse two dispatchers fanning out to the same child with different units at generation time; deterministic unit mapping; tests added.
  - Hardened `cmd/worker` listener: proper readiness gating including flipping not-ready during shutdown, connection timeouts, safe `:0` binding, omits empty `ai:` line, and adds standard security headers.
  - Lint config: own `forbidigo` in `.golangci.yml` and lint `integration`/`livesmoke` tags; gate added to keep this ownership.
  - Split-update marshal failures now return RFC 7807 JSON; removed the stale `STATUS.md` reference and corrected docs.

<sup>Written for commit fe015faeccf2b3e8ffe1a1dc4420b59794d6d5e8. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/gradionhq/margince-poc-v1/pull/457?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added optional worker health, readiness, and metrics endpoints.
  * Added process-local runtime metrics and detailed per-unit job metrics.
  * Readiness checks now report boot, database, and cache status.

* **Bug Fixes**
  * Backfill completion now schedules workspace-scoped digest work.
  * Improved JSON error responses and response replay behavior.
  * Invalid worker settings are rejected earlier with clearer errors.

* **Documentation**
  * Documented observability endpoints, readiness behavior, and expanded metrics.

### CODE REVIEW

File: backend/cmd/worker/config.go [hunk 1]
```diff
@@ -119,6 +108,12 @@ func parseWorkerFlags(args []string) (workerConfig, error) {
 	// verbatim as the River schedule; compose clamps nothing
 	// (compose.WebhookRetryConfig.Interval).
 	fs.DurationVar(&cfg.webhookRetryInterval, "webhook-retry-interval", 30*time.Second, "how often the outbound-webhook retry dispatcher fans one due-retry pass out per live workspace")
+	// Off by default, and off means no listener at all rather than one bound
+	// somewhere harmless: this surface carries workspace ids and has no
+	// authentication of its own, exactly like the api's, so exposing it is an
+	// operator's decision and so is the interface it binds.
+	fs.StringVar(&cfg.observeAddr, "observe-addr", os.Getenv("MARGINCE_OBSERVE_ADDR"),
+		"address to serve this worker's /healthz, /readyz and /metrics on (e.g. 127.0.0.1:9101). Empty serves nothing. Process-local metrics only — the job-table and outbox gauges stay a single fleet-wide reading on the api.")
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/cmd/worker/config.go:117]: Fixed in fe015fa (comment corrected in 22dccf4). You are right and I had already corrected the package doc and the reference docs — this was the sibling copy that got missed, which is exactly the "fixed the case under review, missed the sibling" pattern our rule 1 names. The flag comment now says what is true: no workspace id and no tenant data, but unauthenticated and disclosing dependency health and process capacity, so exposing it and choosing the interface are the operator's.

File: backend/cmd/worker/observe_test.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+// because an operator has to learn to ignore it.
+func TestTheWorkerReadinessBodyOmitsTheLineItWiresNothingFor(t *testing.T) {
+	// A nil pool makes the postgres check fail, so this is the 503 body — the
+	// one an operator reads when something is wrong, and the one that must
+	// name the failed dependency and nothing it does not wire.
+	_, body := get(t, startForTest(t)+"/readyz")
+	if strings.Contains(body, "ai:") {
+		t.Errorf("the worker reports an AI line it wires nothing for: %q", body)
+	}
+}
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/cmd/worker/observe_test.go]: Fixed in 22dccf4. Correct and important — the listener held an incomplete `bootGate`, so the readiness check returned before `Readyz` ever rendered the visibility lines. The assertion was over a 503 body that never had an `ai:` line to omit, so it could not fail. The visibility lines are only reachable on the 200 path, so the assertion moved to the integration lane's real-dependency test where a booted gate, a live pool and a live bus actually produce one.

File: backend/internal/compose/capturejobs_integration_test.go [hunk 3]
```diff
@@ -82,6 +89,66 @@ func TestBackfillCompletionBuildsTheDigest(t *testing.T) {
 	awaitKindCompleted(waitCtx, t, sub, "capture_backfill")
 	// The digest that follows the completed backfill is the payoff wiring.
 	awaitKindCompleted(waitCtx, t, sub, CaptureDigestWorkspaceArgs{}.Kind())
+
+	// The waits above are satisfied by a digest of the right KIND, which the
+	// boot pass also produces — they prove the wiring fires, not WHAT it
+	// fires. What follows is the fix itself, and each half fails silently on a
+	// single-workspace fixture, where a fleet fan-out and a one-off enqueue
+	// produce the same visible outcome.
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/internal/compose/capturejobs_integration_test.go]: Fixed in 22dccf4. "The fix itself" is change history in the source, which our rule 4 rules out. It now states what the assertions distinguish — a one-off workspace child from a fleet dispatcher, which on a single-workspace fixture produce the same visible outcome.

File: backend/internal/compose/jobcensusargs.go [hunk 4]
```diff
@@ -143,6 +143,76 @@ func (c *JobCensus) everyArgsFieldIsDeclaredAndBack() []string {
...
+		return nil, errors.New("the registration recorded no args value at all, so there is nothing to encode")
+	}
+	encoded, err := json.Marshal(args)
+	if err != nil {
+		return nil, fmt.Errorf("its args do not encode at all, so River could never enqueue one: %w", err)
+	}
+	var object map[string]json.RawMessage
+	if err := json.Unmarshal(encoded, &object); err != nil {
+		return nil, fmt.Errorf("its args encode to %s rather than to a JSON object, so river_job.args has no keys to group on", encoded)
+	}
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/internal/compose/jobcensusargs.go:212]: Fixed in 22dccf4. The sibling branch wraps its cause with %w and this one dropped it, so a reader saw the encoded payload without the decoder's reason for rejecting it.

File: backend/jobregistrationban_test.go [hunk 5]
```diff
@@ -210,7 +213,75 @@ func riverForbidRules(t *testing.T) []forbidRule {
...
+		for name, tags := range tagged {
+			if slices.Contains(tags, tag) {
+				covered = append(covered, name)
+			}
+		}
+		if !slices.Contains(covered, ".golangci.yml") {
+			t.Errorf("no repo-wide pass declares the %q build tag, so the %s-only files are linted by nothing that runs tree-wide", tag, tag)
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/jobregistrationban_test.go:282]: Fixed in 22dccf4. Good catch: the test named `.golangci.yml` while the sibling gate only guarantees that *exactly one* config owns forbidigo. If ownership moved, this would have passed about a config that no longer runs the linter. The owner is derived now, and the tag obligation follows it.

File: docs/reference/configuration.md [hunk 6]
```diff
@@ -193,14 +219,25 @@ Four things worth knowing before you build an alert on these:
...
+- **The per-workspace pair reads one grain too coarse for four dispatchers,**
+  and that is what `margince_sweep_units_*` is for. Gmail sync, Gmail watch and
+  the Telegram poll fan out per **connection**; the voice-build retry fans out
+  per **build**. A workspace holding two connections produces two children per
+  pass, and if the broken one failed before the healthy one succeeded, that
+  workspace's most recent child is the successful one — so the workspace pair
+  reports zero failures while a connection is dead. The unit pair counts each
+  connection in its own right and reports the failure. Read the workspace pair
+  for fleet coverage and the unit pair for whether every unit of a pass ran;
+  neither replaces the other, and no kind is in both.
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply docs/reference/configuration.md]: Fixed in 22dccf4. The closing clause contradicted the "overlap rather than partition" paragraph four lines above it. It now says the four finer-grained kinds appear in both, and points at the note on never summing them.

File: backend/internal/compose/capturejobs.go [hunk 7]
```diff
@@ -304,27 +304,43 @@ func (w *captureBackfillWorker) Work(ctx context.Context, job *river.Job[Capture
...
+		w.log.WarnContext(ctx, "capture backfill: no River client in context, so the same-day digest was not enqueued",
+			"backfill", args.BackfillID, "err", err)
 		return
 	}
-	if _, err := client.Insert(ctx, CaptureDigestArgs{}, nil); err != nil {
+	child := CaptureDigestWorkspaceArgs{Workspace: args.Workspace}
+	if _, err := client.Insert(ctx, child, oneOffChildOpts(child.Kind())); err != nil {
 		w.log.WarnContext(ctx, "capture backfill: digest enqueue failed",
-			"backfill", backfillID, "err", err)
+			"backfill", args.BackfillID, "err", err)
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/internal/compose/capturejobs.go:343]: Declining, with the reason rather than silence.

A River retry cannot re-enqueue the digest. `completed` is a one-shot edge: `RunBackfillStep` returns it only when its guarded UPDATE actually moved a live run to `done` (rowsAffected == 1). A retried `capture_backfill` therefore finds the run already finished, gets `done` without `completed`, and never reaches `enqueueDigest` — so returning the error would fail a job whose work is complete and buy nothing.

Nor is this a comment-only waiver. The durable retry is the nightly `capture_digest` dispatcher, a scheduled River job, not a promise in prose: a missed same-day digest costs one morning of freshness and the payload is rebuilt that night (the build replaces the day). What the review is right about is the silent return — the missing-client branch discarded its error, which is indistinguishable from having enqueued. That was fixed in 100084c5; both branches now log.

File: backend/internal/platform/httpserver/observe.go [hunk 8]
```diff
@@ -193,12 +194,24 @@ func Readyz(aiState string, embedState func(context.Context) string, checks ...R
...
+		// embedState and a marker-read that already failed into "unknown" are
+		// deliberately indistinguishable here, so omitting it would turn one
+		// of those two into an absence and the other into a value.
 		embed := "unknown"
 		if embedState != nil {
 			embed = embedState(ctx)
 		}
-		w.WriteHeader(http.StatusOK)
-		_, _ = fmt.Fprintf(w, "ready\nai: %s\nembed: %s\n", aiState, embed)
+		_, _ = fmt.Fprintf(w, "embed: %s\n", embed)
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/internal/platform/httpserver/observe.go:214]: Declining as out of scope, with the reasoning.

Every ignored write named here is pre-existing — `Readyz`'s `_, _ = fmt.Fprintf` and `Metrics`' section calls predate this branch; I changed which visibility lines `Readyz` renders, not how it writes them.

The file also already carries both postures deliberately, and the split is principled rather than accidental: `writeJobMetrics` returns its error because a truncated job section *parses as a smaller fleet*, which an alert would act on. The probe body and the counter sections do not, because a failed write there means the client hung up and there is no channel left to report on — the same rationale the `//craft:ignore swallowed-errors` waivers in this tree state at other response-write sites.

Changing that is a refactor of every section on the shared chassis (`extra`, `writeOverlayMetrics`, the AI counters, `Readyz`), touching roles this PR does not, and it is not a fix to anything this branch introduced. Happy to file it as a follow-up issue if you think the inconsistency is worth closing.

File: backend/internal/platform/jobs/spec.go [hunk 9]
```diff
@@ -40,6 +40,53 @@ const (
...
+// that is where the fan-out decision is made — so a reader holding a child row
+// has to walk the edge backwards to learn what it counts as. This is that walk,
+// done once over the compiled table rather than at every call site that needs
+// it. A child no dispatcher names is absent rather than defaulted: the fan-out
+// edges are the registry of what fans out at all.
+func FanOutUnits() map[string]FanOutUnit {
+	units := map[string]FanOutUnit{}
+	for _, spec := range specs {
+		if spec.FanOutTo != "" {
+			units[spec.FanOutTo] = spec.FanOutUnit
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/internal/platform/jobs/spec.go:89]: Fixed in fe015fa. Real defect and well spotted — the walk was over an unordered map with an overwrite on collision, so with two dispatchers naming one child the winner was whichever the range reached last, and could differ between processes.

Closed at generation rather than papered over at the read: `gen-jobs` now refuses two dispatchers fanning out to one child with *different* units, where a contract can still be edited. Two edges sharing a child with the SAME unit stay legal — the rule is about disagreement, not about sharing, and both cases have a test. `FanOutUnits` also walks in kind order now, so determinism is a property of the function rather than of the file it happens to read.

File: backend/internal/platform/jobs/stats.go [hunk 10]
```diff
@@ -219,3 +282,65 @@ func statsBySweep(ctx context.Context, pool *pgxpool.Pool) ([]SweepPass, error)
...
+		SELECT kind,
+		       count(*)::bigint,
+		       count(*) FILTER (WHERE state IN ` + terminalBadStates + `)::bigint
+		FROM (
+		    SELECT DISTINCT ON (j.kind, j.args->>u.args_key)
+		           j.kind, j.state::text AS state
+		    FROM river_job j
+		    JOIN unnest($1::text[], $2::text[]) AS u(kind, args_key) ON u.kind = j.kind
+		    WHERE '` + SweepTag + `' = ANY(j.tags)
+		      AND coalesce(j.args->>u.args_key, '') <> ''
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/internal/platform/jobs/stats.go:326]: Fixed in fe015fa. Correct: `statsBySweep` excludes a present-but-empty `workspace_id` explicitly, and the unit read did not. Every kind it reads is a tenant kind, so a tagged row with a unit key and no workspace is malformed, and counting it credits a pass with coverage belonging to no tenant. Same predicate, carried over with the reason.

File: backend/cmd/worker/main.go [hunk 11]
```diff
@@ -125,11 +116,81 @@ func run(ctx context.Context, args []string, stdout io.Writer) error {
 		return err
 	}
 	defer stopJobs()
+	// Every phase a replica needs to do work has returned; /readyz may say so.
+	started.complete()
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/cmd/worker/main.go:120]: Fixed in fe015fa. Real gap, and the ordering makes it worse than it looks: the listener is stopped LAST on purpose so the drain stays observable, which meant a terminating replica kept advertising itself for the whole shutdown.

`bootGate` now answers no at both ends of the process's life, and `run()` defers `draining()` AFTER `complete()` so LIFO puts it ahead of the job-runner stop — readiness goes false at the top of the shutdown, not after it. Proven in the integration lane: 200 before, 503 naming the check after.

File: backend/cmd/worker/boothelpers_test.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+//
+// Both cases fail in the flag phase, ahead of the registration. The logger's
+// own refusal — the one phase after it — is driven directly by
+// TestAnUnknownLogLevelOrFormatFailsTheBoot instead.
+func TestConfiguringTheWorkerFailsBeforeAnythingIsOpened(t *testing.T) {
+	for _, tc := range []struct {
+		name, wantIn string
+		args         []string
+	}{
+		{"a missing dsn", "dsn", []string{}},
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/cmd/worker/boothelpers_test.go:41]: Fixed in fe015fa. Right — `--dsn` backs its default with `MARGINCE_DSN`, so a CI exporting one would have flipped that case to a successful boot and the subtest to a failure for a reason unrelated to the code. It sets the variable empty explicitly now; verified by running the test with `MARGINCE_DSN` exported.

File: backend/.golangci.yml [hunk 13]
```diff
@@ -42,6 +66,17 @@ linters:
...
+        # Structured logging is the only logging: a fmt.Print* line carries
+        # no level, no correlation_id, and no fields, so it is invisible to
+        # every reader that joins a log line to an audit row and a bus event.
+        - pattern: '^fmt\.Print.*$'
+          msg: "use slog, not fmt.Print*"
+        # Every refusal this backend writes is an RFC 7807 problem document.
+        # http.Error writes text/plain, so a client that parses the contract's
+        # error shape meets an unparseable body on exactly the paths it needs
+        # most.
+        - pattern: '^http\.Error$'
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/.golangci.yml:78]: Not changing, with the evidence.

The generated `HandlerWithOptions` fallback is already overridden in production: `internal/compose/routes.go:64` wires `ErrorHandlerFunc: paramParseError`, so the `text/plain` path is unreachable on the served surface. The only construction without options is a test (`channelconnect_test.go:28`).

`internal/contracts/api_gen.go` is also generated and drift-frozen — a hand edit fails the drift gate — so the fallback is not this PR's to change either way. The forbidigo rule governs hand-written call sites by design, and the one it caught (the split-update handler) is fixed in this branch.

File: backend/internal/compose/jobmetrics.go [hunk 14]
```diff
@@ -295,6 +300,46 @@ func writeSweepGauges(w io.Writer, sweeps []jobs.SweepPass) error {
...
+	ordered := slices.SortedFunc(slices.Values(units), func(a, b jobs.SweepUnit) int {
+		return cmp.Compare(a.Kind, b.Kind)
+	})
+
+	if err := writeFamilyHeader(w, "margince_sweep_units_total",
+		"Fan-out units with a surviving child of this fleet pass, for the dispatchers that fan out per connection or per build rather than per workspace. The unit label names the grain. Kinds that fan out per workspace are absent here and reported by margince_sweep_workspaces_total, which is the same measurement for them."); err != nil {
+		return err
+	}
+	for _, u := range ordered {
+		if _, err := fmt.Fprintf(w, "margince_sweep_units_total{sweep=%s,unit=%s} %d\n",
```
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply backend/internal/compose/jobmetrics.go:324]: Not changing, with the reasoning.

`margince_sweep_units_total` is the deliberate sibling of `margince_sweep_workspaces_total`, which shipped in #446 with exactly the same shape — a gauge with a `_total` suffix. Renaming one half of two parallel pairs would leave an operator reading `margince_sweep_workspaces_total` beside `margince_sweep_units`, which is a worse surface than the convention the rename would satisfy.

The convention point is fair, and it applies to both families or to neither. Happy to file it as a follow-up to rename the pair together if that is worth doing — it is a breaking change for any existing dashboard, so it wants to be one deliberate move rather than a side effect of this PR.

File: docs/reference/configuration.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (tinnhlt-gra) [reply docs/reference/configuration.md]: Fixed in 22dccf4 — same finding as CodeRabbit's on this line. The closing clause now says the four finer-grained kinds appear in both pairs by design, and points at the note on never summing them.
